### ISSUE TITLE
Stochastic mode to compute transition matrix fails

### ISSUE THREAD

**Issue by Marius1311** [theislab/cellrank#466]
I'm computing a transition matrix using the stochastic mode for the large 48k morris dataset. This works fine using deterministic mode. 

```python
from cellrank.tl.estimators import GPCCA

# set the parameters we want to use to compute the transition matrix
weight_connectivities=0.2
mode="stochastic" 
n_jobs=16
softmax_scale=None

if use_caching:
    c.tl.transition_matrix(adata, 
                           weight_connectivities=weight_connectivities, 
                           mode=mode, 
                           n_jobs=n_jobs, 
                           softmax_scale=softmax_scale, 
                           fname='2021-01-21_transition_matrix_stochastic', 
                           force=force_recompute)
    g_fwd = GPCCA(adata, obsp_key='T_fwd')
else:
    kernel = cr.tl.transition_matrix(adata, 
                                     weight_connectivities=weight_connectivities, 
                                     mode=mode, 
                                     n_jobs=n_jobs, 
                                     softmax_scale=softmax_scale)
    g_fwd = GPCCA(kernel)
    
print(g_fwd)
```

```pytb
No cache found in `2021-01-21_transition_matrix_stochastic.pickle`, computing values.
Computing transition matrix based on logits using `'stochastic'` mode
Estimating `softmax_scale` using `'deterministic'` mode
100%
48515/48515 [00:20<00:00, 26065.98cell/s]
Setting `softmax_scale=2.6275`
0%
0/48515 [00:00<?, ?cell/s]
---------------------------------------------------------------------------
_RemoteTraceback                          Traceback (most recent call last)
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/python/lib/python3.8/site-packages/joblib/externals/loky/process_executor.py", line 431, in _process_worker
    r = call_item()
  File "/opt/python/lib/python3.8/site-packages/joblib/externals/loky/process_executor.py", line 285, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/opt/python/lib/python3.8/site-packages/joblib/_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)
  File "/opt/python/lib/python3.8/site-packages/joblib/parallel.py", line 262, in __call__
    return [func(*args, **kwargs)
  File "/opt/python/lib/python3.8/site-packages/joblib/parallel.py", line 262, in <listcomp>
    return [func(*args, **kwargs)
  File "/opt/python/lib/python3.8/site-packages/cellrank/tl/kernels/_velocity_kernel.py", line 517, in _run_stochastic
    if H.shape == (n_neigh, n_genes):
AttributeError: 'NoneType' object has no attribute 'shape'
"""

The above exception was the direct cause of the following exception:

AttributeError                            Traceback (most recent call last)
<ipython-input-42-ddfe0b493d90> in <module>
      8 
      9 if use_caching:
---> 10     c.tl.transition_matrix(adata, 
     11                            weight_connectivities=weight_connectivities,
     12                            mode=mode,

/storage/groups/ml01/code/marius.lange/scachepy/scachepy/utils.py in __call__(self, *args, **kwargs)
    106 
    107     def __call__(self, *args, **kwargs):
--> 108         return self._fn(*args, **kwargs)
    109 
    110     def __repr__(self):

/storage/groups/ml01/code/marius.lange/scachepy/scachepy/modules.py in wrapper(*args, **kwargs)
    299                     print(f'No cache found in `{str(f) + self.backend.ext}`, ' + ('computing values.' if call else 'searching for values.'))
    300 
--> 301                 res = callback(*args, **kwargs) if call else adata if copy else None
    302                 ret = cache_fn(res if copy else adata, fname, True, False,
    303                                skip, keyhint, watchers, *args, **kwargs)

/opt/python/lib/python3.8/site-packages/cellrank/tl/_transition_matrix.py in transition_matrix(adata, backward, vkey, xkey, gene_subset, mode, backward_mode, scheme, softmax_scale, weight_connectivities, density_normalize, key, **kwargs)
     74         adata, backward=backward, vkey=vkey, xkey=xkey, gene_subset=gene_subset
     75     )
---> 76     vk.compute_transition_matrix(
     77         softmax_scale=softmax_scale,
     78         mode=mode,

/opt/python/lib/python3.8/site-packages/cellrank/tl/kernels/_velocity_kernel.py in compute_transition_matrix(self, mode, backward_mode, scheme, softmax_scale, n_samples, seed, **kwargs)
    293             logg.info(f"Setting `softmax_scale={softmax_scale:.4f}`")
    294 
--> 295         tmat, cmat = _dispatch_computation(
    296             mode,
    297             scheme=scheme,

/opt/python/lib/python3.8/site-packages/cellrank/ul/_utils.py in wrapper(*args, **kwargs)
    272     @wraps(func)
    273     def wrapper(*args, **kwargs):
--> 274         return dispatch(args[0])(*args[1:], **kwargs)
    275 
    276     wrapper.register = register

/opt/python/lib/python3.8/site-packages/cellrank/tl/kernels/_velocity_kernel.py in <lambda>(**kwargs)
    566 
    567 _dispatch_computation.register(VelocityMode.STOCHASTIC)(
--> 568     lambda **kwargs: _run_in_parallel(_run_stochastic, **kwargs)
    569 )
    570 _dispatch_computation.register(VelocityMode.DETERMINISTIC)(

/opt/python/lib/python3.8/site-packages/cellrank/tl/kernels/_velocity_kernel.py in _run_in_parallel(fn, conn, **kwargs)
    358     kwargs["indptr"] = conn.indptr
    359 
--> 360     return parallelize(
    361         fn,
    362         ixs,

/opt/python/lib/python3.8/site-packages/cellrank/ul/_parallelize.py in wrapper(*args, **kwargs)
     99             pbar, queue, thread = None, None, None
    100 
--> 101         res = jl.Parallel(n_jobs=n_jobs, backend=backend)(
    102             jl.delayed(callback)(
    103                 *((i, cs) if use_ixs else (cs,)),

/opt/python/lib/python3.8/site-packages/joblib/parallel.py in __call__(self, iterable)
   1052 
   1053             with self._backend.retrieval_context():
-> 1054                 self.retrieve()
   1055             # Make sure that we get a last message telling us we are done
   1056             elapsed_time = time.time() - self._start_time

/opt/python/lib/python3.8/site-packages/joblib/parallel.py in retrieve(self)
    931             try:
    932                 if getattr(self._backend, 'supports_timeout', False):
--> 933                     self._output.extend(job.get(timeout=self.timeout))
    934                 else:
    935                     self._output.extend(job.get())

/opt/python/lib/python3.8/site-packages/joblib/_parallel_backends.py in wrap_future_result(future, timeout)
    540         AsyncResults.get from multiprocessing."""
    541         try:
--> 542             return future.result(timeout=timeout)
    543         except CfTimeoutError as e:
    544             raise TimeoutError from e

/opt/python/lib/python3.8/concurrent/futures/_base.py in result(self, timeout)
    437                 raise CancelledError()
    438             elif self._state == FINISHED:
--> 439                 return self.__get_result()
    440             else:
    441                 raise TimeoutError()

/opt/python/lib/python3.8/concurrent/futures/_base.py in __get_result(self)
    386     def __get_result(self):
    387         if self._exception:
--> 388             raise self._exception
    389         else:
    390             return self._result

AttributeError: 'NoneType' object has no attribute 'shape'
```
Basically, computing the softmax scale works fine, but as soon as it starts to go into stochastic mode, it fails. 

#### Versions:

cellrank==1.1.0 scanpy==1.6.1 anndata==0.7.5 numpy==1.19.5 numba==0.52.0 scipy==1.6.0 pandas==1.2.0 scikit-learn==0.24.0 statsmodels==0.12.1 python-igraph==0.8.3 scvelo==0.2.2 pygam==0.8.0 matplotlib==3.3.3 seaborn==0.11.1

**Comment by michalk8** [comment]
@Marius1311 try
```pip install jax jaxlib``, it's most likely not in your Python env. Not sure why it allows you to do this though.

**Comment by Marius1311** [comment]
Thanks! I'm just trying this, let's see what happens.

**Comment by Marius1311** [comment]
worked, thanks.

**Comment by EmilyMiddendorp** [comment]
Hi! 
I saw this post and I have a similar problem. I am trying to run a 88k cells dataset in stochastic mode but by the second step (after calculating de softmax_scale), the code gets a RAM peak and kill my kernel. I checked if tax and jaxlib are on my environment and they are. I also tried to load a smaller dataset (±42k cells) but the kernel still died after a RAM peak. I am using n_jobs=1 to serialise the process hoping that would help but the code still takes all of my cores anyway. Any idea on how I can solve this?
